liota/dccs/iotcc.py

```diff
@@ -398,28 +401,37 @@ def set_properties(self, reg_entity_obj, properties):
             # get dev_type, and prop_dict if possible
             with self.file_ops_lock:
                 self._store_reg_entity_attributes("Devices", entity, reg_entity_obj.reg_entity_id,
-                                                 entity.entity_type, properties)
+                                                  entity.entity_type, properties)
 
     def publish_unit(self, reg_entity_obj, metric_name, unit):
         """
          Publish SI units as properties for Metrics but RegisteredMetric object are simply returned
          so currently units are set to the parent RegisteredEntity(Device or EdgeSystem)
+         Either units are set with prefix as properties or only unit gets set as property if prefix doesn't exist.
         :param reg_entity_obj: RegisteredEntity Object
         :param metric_name: Metric Name
         :param unit: SI Unit
         :return:
         """
         str_prefix, str_unit_name = parse_unit(unit)
-        if not isinstance(str_prefix, basestring):
-            str_prefix = ""
-        if not isinstance(str_unit_name, basestring):
-            str_unit_name = ""
-        properties_added = {
-            metric_name + "_unit": str_unit_name,
-            metric_name + "_prefix": str_prefix
-        }
-        self.set_properties(reg_entity_obj, properties_added)
-        log.info("Published metric unit with prefix to IoTCC")
+        if not isinstance(str_prefix, basestring) and isinstance(str_unit_name, basestring):
+            properties_unit_list = {
+                metric_name + "_unit": str_unit_name
+            }
+            log.debug("Publishing unit {0} for metric {1} to IoTCC".format(str_unit_name, metric_name))
+        elif isinstance(str_unit_name, basestring) and isinstance(str_prefix, basestring):
+            properties_unit_list = {
+                metric_name + "_unit": str_unit_name,
+                metric_name + "_prefix": str_prefix
+            }
+            log.debug("Publishing unit {0} with prefix {1} for metric {2} to IoTCC".format(str_unit_name, str_prefix,
+                                                                                           metric_name))
+        else:
+            properties_unit_list = {}
+            log.debug("{0} metric unit with prefix cannot be parsed and published to IoTCC".format(metric_name))
```

Comment:
Is this the only situation when the code will reach else part? can it not reach here when no units are set?

I think the if..elif..else part is a bit confusing...can we change it to:

si_properties = {}
if (isinstance(str_unit_name, basestring):
          si_properties[metric_name + "_unit"]: str_unit_name
          if isinstance(str_prefix, basestring):
                   si_properties[metric_name + "_prefix"]: str_prefix
 if si_properties:
         self.set_properties()


What do you think?

Reply:
Thanks for the suggestion. But, I believe elif provides the clear explanation of all the three use cases and the better understanding of the three scenarios we can face with units and prefix.

Reply:
As far as i understand, the else part will be visited even if no unit is set. The debug log in the else part does not indicate that. Please correct me if i am wrong.

Reply:
if entity_obj.unit is not None:                
self.publish_unit(reg_entity_child, entity_obj.name, entity_obj.unit)

If unit exists then only publish_unit method will be called.

Reply:
ok, understood. Should publish_unit be an internal method (with an underscore) in that case?

Reply:
yes it should be there are multiple methods which are required to be refactored in the similar way in the Liota codebase. And I believe that should be handled in separate PR as they are previously introduced methods by various developers.

Reply:
I agree we should change it in whole code base later. However, iotcc.py is the primary module we will be exposing to liota package writers and hence, having public methods in this class will lead to unnecessary usage of these methods. I see only two methods in this class which are public but not part of the interface (get_properties and publish_unit: Hopefully i didnt miss any). Hence i think it will be better to at least make this method private. (get_properties might still be debatable and introduced as public knowingly.)

Reply:
As, I mentioned above that entire change should be handled in separate PR and it should be marked as separate issue to revisit all methods of IoTCC. This PR is meant only for fixing the unit,prefix code changes.

Reply:
Sure, I understand. Let's take it up later then.